Do not return all indices if a specific alias is requested via get aliases api.

martijnvg opened the pull request:
If a get alias api call requests a specific alias pattern then
indices not having any matching aliases should not be included in the response.

This is a second attempt to fix this (first attempt was #28294).
The reason that the first attempt was reverted is because when xpack
security is enabled then index expression (like * or _all) are resolved
prior to when a request is processed in the get aliases transport action,
then `MetaData#findAliases` can't know whether requested all where
requested since it was already expanded in concrete alias names. This
change now adds an additional field to the request calss that keeps track
whether all aliases where requested.

Closes #27763

elasticmachine commented:
Pinging @elastic/es-core-infra

martijnvg commented on server/src/main/java/org/elasticsearch/action/admin/indices/alias/get/GetAliasesRequest.java:
```diff
@@ -107,4 +121,17 @@ public ActionRequestValidationException validate() {
         throw new UnsupportedOperationException("usage of Streamable is to be replaced by Writeable");
     }
+
+    private static boolean matchAllAliases(final String[] aliases) {
```
The tests in #25114 use wildcard in index names and that is why I think no test failed. But I think it makes sense to also return `true` if `*` is used as alias name.

martijnvg commented:
@olcbean I agree, this is a tricky change.

> new GetAliasesRequest().aliases("alias") will have matchAllAliases=true. Could such request return more results that intended ? new GetAliasesRequest("alias"); / new GetAliasesRequest().setAliases("alias"); will work as expected.

`aliases(...)` setter is used in xpack security to overwrite aliases that a user is allowed to use based on security constraints. It is also by the java clients as a setter. I would prefer this setter to be named something else to describe this use case too. That is why I added an additional setter method. When xpack security replaces aliases (and indices) these values are already resolved, so the get aliases api does not know what the user originally specified. That is why I keep track of whether a match all syntax is used in a boolean field.

I'm not happy about this change either, but it is the best I came up with to fix #27763

javanna commented on server/src/main/java/org/elasticsearch/action/admin/indices/alias/get/GetAliasesRequest.java:
```diff
@@ -88,6 +92,16 @@ public GetAliasesRequest indicesOptions(IndicesOptions indicesOptions) {
 
+    public GetAliasesRequest setAliases(String... aliases) {
+        this.aliases = aliases;
+        this.matchAllAliases = matchAllAliases(aliases);
```
As far as I understand in the REST layer, we don't print out any index for which there are no aliases to return, but only in case the alias (name) parameter was provided.(https://github.com/elastic/elasticsearch/blob/master/server/src/main/java/org/elasticsearch/rest/action/admin/indices/RestGetAliasesAction.java#L139). I believe this change tries to mimic the same behaviour for the transport client, but in the REST layer we don't look at what the name parameter matches, only at whether it's provided or not:

```
curl localhost:9200/_alias?pretty
{
  "index2" : {
    "aliases" : { }
  },
  "index" : {
    "aliases" : {
      "alias" : { }
    }
  }
}

curl localhost:9200/_alias/_all?pretty
{
  "index" : {
    "aliases" : {
      "alias" : { }
    }
  }
}
```

That may be right or wrong, but I think we should try to return the same results if we make this change, so I'd say without changing the behaviour at REST (although we may want to discuss what the right thing to do is) the best we can do at transport is to only look at whether any specific alias was requested rather than whether the expression matched all or not.

Furthermore, I'd expect that once these changes are made to `MetaData`, the REST action should be updated as some logic can be removed? By the way, related but it should not affect this PR, there's also #28799 under review that is moving the REST logic to the transport layer, yet the REST logic remains in `GetAliasesResponse#toXContent` which doesn't change what the transport client returns.

I would probably consider renaming the `AliasesRequest#aliases` method (to `replaceAliases`?) and make sure that it's only used internally (although it needs to be public), and have users call the current setter. We clearly can't have both call the same method or we lose information about what was set in the first place. That would make it possible to keep track of whether specific aliases were requested or not through a flag, similarly to what you do now.

martijnvg replied:
> That may be right or wrong, but I think we should try to return the same results if we make this change, so I'd say without changing the behaviour at REST (although we may want to discuss what the right thing to do is) the best we can do at transport is to only look at whether any specific alias was requested rather than whether the expression matched all or not.

Makes sense. I will make sure that the transport client doesn't return indices with no aliases if `_all` is provided.

> Furthermore, I'd expect that once these changes are made to MetaData, the REST action should be updated as some logic can be removed? By the way, related but it should not affect this PR, there's also #28799 under review that is moving the REST logic to the transport layer, yet the REST logic remains in GetAliasesResponse#toXContent which doesn't change what the transport client returns.

Agreed. I will wait for #28799 to get in before working on that.

> I would probably consider renaming the AliasesRequest#aliases method (to replaceAliases?) and make sure that it's only used internally (although it needs to be public), and have users call the current setter. We clearly can't have both call the same method or we lose information about what was set in the first place. That would make it possible to keep track of whether specific aliases were requested or not through a flag, similarly to what you do now.

Yes, I totally agree here.

martijnvg commented:
@olcbean Sorry for the late reply.

> Can you modify the x-pack call ?

Yes, that is what @javanna suggested too.

> can add an explicit matchAllAliases setter instead of implicitly detecting it depending on which aliases setter has been used 

I think we shouldn't do that otherwise users of the Java client and we in the rest action may forget to set it.

martijnvg commented:
@javanna I've updated the PR and addressed your first feedback point.

Regarding the other feedback points:
* Cleanup `RestGetAliasesAction`
* Renaming the AliasesRequest#aliases(...) method to AliasesRequest#replaceAliases(...)

I think we should that in follow up PRs in order to make the changes digestible. As the PR it now it should fix the problem in #27763.

javanna commented on server/src/main/java/org/elasticsearch/cluster/metadata/MetaData.java:
```diff
@@ -256,44 +256,52 @@ public boolean equalsAliases(MetaData other) {
             for (ObjectCursor<AliasMetaData> cursor : indexMetaData.getAliases().values()) {
                 AliasMetaData value = cursor.value;
-                if (matchAllAliases || Regex.simpleMatch(aliases, value.alias())) {
+                if (Strings.isAllOrWildcard(aliases) || Regex.simpleMatch(aliases, value.alias())) {
```
this is not exactly the same as the previous matchAllAliases method. I always wondered why, so maybe a good time to fix it, but this will change the behaviour in subtle ways when a '*' is provided or when an array the contains '_all' with more than one item is provided..

martijnvg replied:
> but this will change the behaviour in subtle ways when a '*' is provided or when an array the contains '_all' with more than one item is provided

If _all, * with other alias names is provided then `Strings.isAllOrWildcard(...)` return false, so I think that is ok? I will also look into if it is possible to change TransportGetAliasesAction so that we don't have to change this method, like I mentioned in the other comment.

javanna replied:
but the previous matchAllAliases didn't look for the wildcard, and only looked at whether _all was anywhere within the array.

martijnvg replied:
> and only looked at whether _all was anywhere within the array

Or whether the array was an empty array, which sort of matched with the logic inside `RestGetAliasesAction`.

If you take a look at `RestGetAliasesAction` at line 81 where `paramAsStringArrayOrEmptyIfAll(...)` gets execute, it will replace `_all` and `*` with an empty array.

javanna replied:
good point. ok then we can say that we are making transport client work the same way as REST, that should be good

javanna commented on server/src/main/java/org/elasticsearch/cluster/metadata/MetaData.java:
```diff
@@ -256,44 +256,52 @@ public boolean equalsAliases(MetaData other) {
+                mapBuilder.put(index, Collections.unmodifiableList(filteredValues));
+            } else if (aliasesProvided == false) {
+                // in case all aliases are requested then it is desired to return the concrete index with no aliases (#25114):
+                mapBuilder.put(index, Collections.emptyList());
```
given how weird this API is, I wonder if we could keep this logic contained in TransportGetAliasesAction, similar to what we have in RestGetAliasesAction, rather than introducing this flag in MetaData and exposing it to potentially other API.

martijnvg replied:
Makes sense. I will try to move it to TransportGetAliasesAction.

javanna commented on server/src/main/java/org/elasticsearch/action/admin/indices/alias/get/GetAliasesRequest.java:
```diff
@@ -88,6 +93,20 @@ public GetAliasesRequest indicesOptions(IndicesOptions indicesOptions) {
         return aliases;
     }
 
+    public GetAliasesRequest setAliases(String... aliases) {
```
I don't think this is enough for transport client users. We are trying to fix the transport client behaviour, but effectively for this to work, transport client users need to migrate to the new setter, while the previous "setter" method is still around. How will users know that they have to move away from it? I would  rather move our own code (x-pack) to use a different method to replace aliases, and leave the original setter alone, but have it set the flag that we need. Would that be possible?

martijnvg replied:
I was afraid that this change would be much larger compared to what is currently done in the PR. But taking a second look at it, I think is is ok. I will rename the existing aliases(...) setter to replaceAliases(...)

javanna replied:
great!

olcbean commented on server/src/main/java/org/elasticsearch/action/admin/indices/alias/get/GetAliasesRequest.java:
```diff
@@ -27,20 +28,18 @@
     private String[] aliases = Strings.EMPTY_ARRAY;
-
     private IndicesOptions indicesOptions = IndicesOptions.strictExpand();
+    private boolean aliasesProvided = false;
```
I keep on having problems remembering what `aliasesProvided` means as with either `aliases` or `setAliases` aliases can be specified ...
What about adding some comments to the `aliases` and `setAliases` to make the difference more obvious to the users of this class ?

olcbean commented on server/src/main/java/org/elasticsearch/action/admin/indices/alias/get/GetAliasesRequest.java:
```diff
@@ -27,20 +28,18 @@
-    public GetAliasesRequest(String alias) {
-        this.aliases = new String[]{alias};
+    public GetAliasesRequest(String... aliases) {
+        this.aliases = Objects.requireNonNull(aliases);
```
Can this lead to user code change? ( as you changed the tests above )

martijnvg commented:
> I would really like to avoid any possible code change on the side of the transport client users.

Yes, that is what @javanna wants too. The idea is that changing the `AliasesRequest.aliases(...)` method to`AliasesRequest#replaceAliases(...)` and keeping the original aliases(...) setter on the subclasses will mean no changes to the transport client users. This will address most of your comments.

martijnvg commented:
@javanna I've updated the PR. The original aliases are now include in `GetAliasesRequest`, which may help to fix #31516 too.

javanna commented on server/src/main/java/org/elasticsearch/action/admin/indices/alias/get/TransportGetAliasesAction.java:
```diff
@@ -63,6 +67,17 @@ protected GetAliasesResponse newResponse() {
+        // in case all aliases are requested then it is desired to return the concrete index with no aliases (#25114):
+        boolean aliasesProvided = Strings.isAllOrWildcard(request.getOriginalAliases());
+        ImmutableOpenMap.Builder<String, List<AliasMetaData>> mapBuilder = ImmutableOpenMap.builder(result);
+        Iterable<String> intersection = HppcMaps.intersection(ObjectHashSet.from(concreteIndices), state.metaData().indices().keys());
```
I am not sure why we need the intersection here. We have just resolved the indices to concrete indices based on the same cluster state, so it should all be existing indices?

javanna commented on server/src/main/java/org/elasticsearch/action/admin/indices/alias/get/TransportGetAliasesAction.java:
```diff
@@ -63,6 +67,17 @@ protected GetAliasesResponse newResponse() {
+        Iterable<String> intersection = HppcMaps.intersection(ObjectHashSet.from(concreteIndices), state.metaData().indices().keys());
+        for (String index : intersection) {
+            if (result.get(index) == null && aliasesProvided) {
+                mapBuilder.put(index, Collections.emptyList());
```
do we want to assert that the index was not already there, just to make sure?

martijnvg commented:
@javanna Thanks for reviewing. I've updated the PR.
